File: app/android/app/src/main/java/io/github/nguyenquocanhz/termward/TermwardCorePlugin.java
package io.github.nguyenquocanhz.termward;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;

import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;

import io.github.nguyenquocanhz.termward.mobile.Mobile;

/** Bridge used by app/src/lib/mobile.ts. */
@CapacitorPlugin(
        name = "TermwardCore",
        permissions = {@Permission(strings = {Manifest.permission.POST_NOTIFICATIONS}, alias = "notifications")})
public class TermwardCorePlugin extends Plugin {
    static TermwardCorePlugin instance;
    private static final String BG = "background";

    @Override
    public void load() {
        instance = this;
        if (prefs().getBoolean(BG, false)) startMonitor();
        handleIntent(getActivity().getIntent());
    }

    @PluginMethod
    public void start(PluginCall call) {
        String language = call.getString("language", "en");
        new Thread(() -> {
            try {
                TermwardCore.ensureStarted(getContext(), language);
                JSObject r = new JSObject();
                r.put("port", TermwardCore.port());
                r.put("token", TermwardCore.token());
                call.resolve(r);
            } catch (Exception e) {
                call.reject("Termward core failed to start: " + e.getMessage(), e);
            }
        }).start();
    }

    @PluginMethod
    public void setLanguage(PluginCall call) {
        Mobile.setLanguage(call.getString("language", "en"));
        call.resolve();
    }

    @PluginMethod
    public void getBackground(PluginCall call) {
        JSObject r = new JSObject();
        r.put("enabled", prefs().getBoolean(BG, false));
        call.resolve(r);
    }

    @PluginMethod
    public void setBackground(PluginCall call) {
        boolean on = Boolean.TRUE.equals(call.getBoolean("enabled", false));
        prefs().edit().putBoolean(BG, on).apply();
        if (on) startMonitor();
        else getContext().stopService(new Intent(getContext(), MonitorService.class));
        call.resolve();
    }

    /** Called for launches and re-launches from an alert notification. */
    void handleIntent(Intent intent) {
        if (intent == null) return;
        String hostId = intent.getStringExtra("hostId");
        if (hostId == null) return;
        intent.removeExtra("hostId");
        JSObject d = new JSObject();
        d.put("hostId", hostId);
        notifyListeners("notificationTap", d, true);
    }

    private void startMonitor() {
        ContextCompat.startForegroundService(getContext(), new Intent(getContext(), MonitorService.class));
    }

    private SharedPreferences prefs() {
        return getContext().getSharedPreferences("termward", Context.MODE_PRIVATE);
    }
}

File: app/native/ios-core/Sources/TermwardCorePlugin/TermwardCorePlugin.swift
import Capacitor
import Foundation
import Security
import Termwardcore
import UserNotifications

/// iOS bridge used by app/src/lib/mobile.ts: starts the Go core inside the app
/// and posts its health alerts as local notifications.
@objc(TermwardCorePlugin)
public class TermwardCorePlugin: CAPPlugin, CAPBridgedPlugin, NotificationHandlerProtocol {
    public let identifier = "TermwardCorePlugin"
    public let jsName = "TermwardCore"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setLanguage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBackground", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setBackground", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
    ]

    private static var port = 0
    private static var token = ""
    private let notifier = AlertNotifier()

    override public func load() {
        // Receive taps on our alert notifications.
        bridge?.notificationRouter.localNotificationHandler = self
    }

    @objc func start(_ call: CAPPluginCall) {
        let language = call.getString("language") ?? "en"
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try TermwardCorePlugin.ensureStarted(language: language, notifier: self.notifier)
                call.resolve(["port": TermwardCorePlugin.port, "token": TermwardCorePlugin.token])
            } catch {
                call.reject("Termward core failed to start: \(error.localizedDescription)")
            }
        }
    }

    private static func ensureStarted(language: String, notifier: AlertNotifier) throws {
        if port != 0 { return }
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("core", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        // Keep the data out of iCloud/iTunes backups: it holds private keys.
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        var d = dir
        try? d.setResourceValues(values)

        var p = 0
        try MobileStart(dir.path, try VaultKey.get(), language, notifier, &p)
        port = p
        token = MobileToken()
    }

    @objc func setLanguage(_ call: CAPPluginCall) {
        MobileSetLanguage(call.getString("language") ?? "en")
        call.resolve()
    }

    // iOS suspends apps in the background, so there is no background mode.
    @objc func getBackground(_ call: CAPPluginCall) {
        call.resolve(["enabled": false])
    }

    @objc func setBackground(_ call: CAPPluginCall) {
        call.reject("Background monitoring is not available on iOS")
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { granted, _ in
            call.resolve(["notifications": granted ? "granted" : "denied"])
        }
    }

    // MARK: NotificationHandlerProtocol

    public func willPresent(notification: UNNotification) -> UNNotificationPresentationOptions {
        return [.banner, .list, .sound]
    }

    public func didReceive(response: UNNotificationResponse) {
        if let hostId = response.notification.request.content.userInfo["hostId"] as? String {
            notifyListeners("notificationTap", data: ["hostId": hostId], retainUntilConsumed: true)
        }
    }
}

/// Called by the Go core for every health alert.
final class AlertNotifier: NSObject, MobileNotifierProtocol {
    func post(_ title: String?, body: String?, hostID: String?) {
        let content = UNMutableNotificationContent()
        content.title = title ?? "Termward"
        content.body = body ?? ""
        content.sound = .default
        content.userInfo = ["hostId": hostID ?? ""]
        // One notification per server: a newer state replaces the older one.
        let request = UNNotificationRequest(identifier: "host-\(hostID ?? "")", content: content, trigger: nil)
        UNUserNotificationCenter.current().add(request)
    }
}

/// 32-byte key that encrypts remembered passwords, kept in the Keychain on
/// this device only (never synced).
enum VaultKey {
    private static let service = "io.github.nguyenquocanhz.termward"
    private static let account = "vault"

    static func get() throws -> String {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
        ]
        var out: AnyObject?
        if SecItemCopyMatching(query as CFDictionary, &out) == errSecSuccess, let data = out as? Data {
            return data.map { String(format: "%02x", $0) }.joined()
        }
        var bytes = [UInt8](repeating: 0, count: 32)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
            throw NSError(domain: "Termward", code: 1, userInfo: [NSLocalizedDescriptionKey: "no randomness"])
        }
        let add: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            kSecValueData as String: Data(bytes),
        ]
        let status = SecItemAdd(add as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw NSError(domain: "Termward", code: Int(status), userInfo: [NSLocalizedDescriptionKey: "keychain error \(status)"])
        }
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}
